packages/container/libraries/strongly-typed/src/inject.ts

```diff
@@ -0,0 +1,24 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import type { BindingMap } from './container';
+
+type TypedDecorator<T> = <TTarget, TKey, TIndex>(
+  target: TTarget extends new (...args: any[]) => any
+    ? TTarget
+    : TKey extends keyof TTarget
+      ? Record<TKey, T>
+      : never,
+  key: TKey extends keyof TTarget ? TKey : PropertyKey | undefined,
+  indexOrPropertyDescriptor?: TTarget extends new (...args: infer P) => any
+    ? TIndex extends keyof P
+      ? P[TIndex] extends T
+        ? TIndex
+        : never
+      : never
+    : unknown,
+) => void;
+
+export type TypedInject<TBindingMap extends BindingMap> = <
+  TKey extends keyof TBindingMap,
+>(
+  identifier: TKey,
+) => TypedDecorator<TBindingMap[TKey]>;
```

Comment:
```suggestion
) => TypedDecorator<Awaited<TBindingMap[TKey]>>;
```

Container injects the awaited value of a promise.

Reply:
(I don't think I knew that was how this worked 😅 I think I always assumed `@inject()` needed synchronous dependencies)